design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,433 @@
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
+- support for the Filestore backing store
+- support for devices not provisioned with `ceph-volume`
+
+
+## New OSD operator architecture
+The new OSD operator design focuses only on devices provisioned with Bluestore using `ceph-volume`.
+Some design thought is given to allowing space for support of post-Bluestore backing stores.
+
+### Design considerations
+#### Ceph's Rook orchestrator design
+While the Ceph community has been planning the design of the Ceph orchestrator interface and of
+Ceph's Rook orchestrator, it has remained clear that there are two distinct usage patterns in Rook
+that are important to plan for. The first and most currently most common usage pattern is that of a
+Kubernetes user/admin who wants a simple, hands-off means of creating Ceph storage for their
+Kubernetes cluster/applications. The second usage pattern, is that of an administrator who needs
+more granular control over the operation of their Ceph storage cluster.
+
+While both Rook and Ceph play a large part in creating a good user experience for both usage
+patterns, these can be simplified into **Rook-managed** and **Ceph-managed** categories
+respectively. In the **Rook-managed** pattern, Rook will continue to offer simplified management
+operations in the same manner as Rook does today, abstracting some of the smaller details from the
+user. In the **Ceph-managed** pattern, crucial for users with strict performance requirements of any
+kind, the user will use Ceph's CLI and/or dashboard to manage Ceph daemons and data pools
+explicitly. This will look more familiar to the way existing Ceph clusters are managed today.
+
+##### Rook-managed mode
+Rook-managed mode is further broken down into two sub-modes, **Automatic Rook-managed mode** and
+**Declarative Rook-managed mode**.
+
+###### Automatic Rook-managed mode
+This mode is characterized by either of the following Ceph cluster configuration options:
+`useAllDevices: true`, or`deviceFilter: <set>`. This means that every applicable
+unused device is turned into an OSD, on whichever nodes are appropriate (either specifically-labeled
+or `useAllHosts: true`). In addition, on any host that is available to the Rook cluster, any device
+labeled with the Ceph cluster UUID and corresponds to a valid OSD in the cluster's OSDMap should be
+started.
+
+Rook may opt to delete OSDs which have been `out` for longer than some threshold period measured in
+days, e.g., 45 days. Rook will occasionally report in its logs (at least once per day) that it intends to delete an OSD
+after some number of days.
+
+###### Declarative Rook-managed mode
+This mode is characterized by the following Ceph cluster configuration options:
+`useAllDevices: false`, `deviceFilter: <unset>`, and `devices: <set>`. This means that only devices
```

Comment:
To me `deviceFilter` and `devices` could be merged into a single entity. I know it's a matter of variable *type* but it'd nice.
Also, I believe `deviceFilter` is a form of declarative.

Reply:
I disagree here that `deviceFilter` is declarative. At least what I mean with declarative here is that each disk is declared, and disks could potentially be removed via omission. I suppose the same is possibly true with `deviceFilter`, but I can't imagine anyone wanting to use it that way in reality. To me `deviceFilter` is an extension of the idea "use all of the disks available" the same as `useAllDevices`, but focused on eliminating unwanted disks, resulting in a subset, "use all disks found in this subset." I could foresee a common device filter being `/dev/sd[c-z][a-z]*` as an example, and that is much closer to `useAllDevices` than to specifying devices individually.